### ISSUE TITLE
[library] panic when exported function contains a syntax error

### ISSUE THREAD

**Issue by cppforlife** [vmware-tanzu/carvel-ytt#156]
https://get-ytt.io/#gist:https://gist.github.com/cppforlife/b9d54ff11278a4e1cfe5d2c8c96f62ca

```
error: panic: Expected to find compiled template: Expected to find '_ytt_lib/app/config.lib.yml' compiled template

goroutine 1 [running]:
github.com/k14s/ytt/pkg/template.CompiledTemplateMultiError.buildPos(0x0, 0x0, 0x0, 0xb11560, 0xc000092a00, 0xc000132940, 0x1a00000006, 0x0, 0x0, 0x0, ...)
	github.com/k14s/ytt@/pkg/template/compiled_template_error.go:157 +0x467
github.com/k14s/ytt/pkg/template.CompiledTemplateMultiError.buildEvalErr(0x0, 0x0, 0x0, 0xb11560, 0xc000092a00, 0xc00013a9f0, 0x0, 0x0, 0x0, 0x0, ...)
	github.com/k14s/ytt@/pkg/template/compiled_template_error.go:140 +0x105
github.com/k14s/ytt/pkg/template.NewCompiledTemplateMultiError(0xb0d4c0, 0xc00013a9f0, 0xb11560, 0xc000092a00, 0x0, 0x0)
	github.com/k14s/ytt@/pkg/template/compiled_template_error.go:53 +0x7e4
github.com/k14s/ytt/pkg/template.(*CompiledTemplate).Eval(0xc000064b40, 0xc000092cd0, 0xb11560, 0xc000092a00, 0xc000092640, 0xc000092cd0, 0xc0000fb740, 0x20, 0x8fcda0)
	github.com/k14s/ytt@/pkg/template/compiled_template.go:120 +0x70e
github.com/k14s/ytt/pkg/workspace.(*TemplateLoader).EvalYAML(0xc000092a00, 0xc000092690, 0xc000092690, 0xc000092640, 0xc00000e1a8, 0x1, 0x1, 0xc00009f650)
	github.com/k14s/ytt@/pkg/workspace/template_loader.go:171 +0x686
github.com/k14s/ytt/pkg/workspace.(*LibraryLoader).eval(0xc00014ba68, 0xc00001f040, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
	github.com/k14s/ytt@/pkg/workspace/library_loader.go:140 +0xac7
github.com/k14s/ytt/pkg/workspace.(*LibraryLoader).Eval(0xc00014ba68, 0xc00001f040, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0)
	github.com/k14s/ytt@/pkg/workspace/library_loader.go:91 +0x77
github.com/k14s/ytt/pkg/cmd/template.(*TemplateOptions).RunWithFiles(0xc000097540, 0xc0000317c0, 0x2, 0x2, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
	github.com/k14s/ytt@/pkg/cmd/template/cmd.go:132 +0x4b5
github.com/k14s/ytt/pkg/cmd/template.(*TemplateOptions).Run(0xc000097540, 0x0, 0x0)
	github.com/k14s/ytt@/pkg/cmd/template/cmd.go:85 +0x28b
github.com/k14s/ytt/pkg/cmd/template.NewCmd.func1(0xc0000b6dc0, 0xc0000fae10, 0x0, 0x3, 0x0, 0x0)
	github.com/k14s/ytt@/pkg/cmd/template/cmd.go:53 +0x2a
github.com/cppforlife/cobrautil.WrapRunEForCmd.func1.1(0xc0000b6dc0, 0xc0000fae10, 0x0, 0x3, 0x0, 0x0)
...
```

**Comment by pivotaljohn** [comment]
@GrahamDumpleton's report ☝🏻  contains additional valuable nuance.

**Comment by pivotaljohn** [comment]
Diagnosis (in progress):

- The `CompiledTemplate` (or more precisely, the `CompiledTemplateMultiError` sub-component) assumes that whatever source of Starlark that has errored out is available as a template to the `TemplateLoader` that was driving the evaluation.
  - `CompiledTemplateMultiError` is attempting to build up the error message _including filename, position, and source text_ of the implicated line of Starlark source.
- In certain cases — like when the erroring Starlark is code exported from another library — that Starlark source (referred to internally as a "template") was not registered as a "template" with the `TemplateLoader`.
  - this will also be the case for any generated Starlark source.

_(in progress; will update this comment as a summary as more details are discovered.)_

**Comment by pivotaljohn** [comment]
Published in [v0.42.0](https://github.com/vmware-tanzu/carvel-ytt/releases/tag/v0.42.0).
